### PR TITLE
Re-open stderr to /dev/null on -2 flag

### LINKED ISSUE #8894: Segmentation fault with -2 flag
Hello,

I'm here via https://github.com/radare/radare2-r2pipe/issues/26 (disable warnings from r2pipe by using the -2 ("close stderr") flag).

However, using commit 754e0e32f9aef631d204e8f441c677143c5cf387 with -2 on `/bin/ls` on Ubuntu 16.04.3 crashes: something in the python bindings is not happy. Python bindings were built/installed from the same commit.

    martin@ubuntu14:~/radare2$ r2 /bin/ls
    [0x004049a0]> quit
    martin@ubuntu14:~/radare2$ r2 -2 /bin/ls
    Segmentation fault (core dumped)
    martin@ubuntu14:~/radare2$ gdb /home/martin/bin/prefix/radare2/bin/radare2 core
    Reading symbols from /home/martin/bin/prefix/radare2/bin/radare2...done.
    [New LWP 28962]
    [Thread debugging using libthread_db enabled]
    Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
    
    warning: the debug information found in "/lib64/ld-2.23.so" does not match "/lib64/ld-linux-x86-64.so.2" (CRC mismatch).
    
    Core was generated by `radare2 -2 /bin/ls'.
    Program terminated with signal SIGSEGV, Segmentation fault.
    #0  0x00007f8762ffb1ac in PyObject_Call ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    => 0x00007f8762ffb1ac <PyObject_Call+12>:	48 8b 98 80 00 00 00	mov    0x80(%rax),%rbx
    (gdb) bt
    #0  0x00007f8762ffb1ac in PyObject_Call ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #1  0x00007f8762ffbf9f in _PyObject_CallFunction_SizeT ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #2  0x00007f8763041faf in ?? () from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #3  0x00007f8763042320 in PyFile_FromFile ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #4  0x00007f8763006a2b in _PySys_Init ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #5  0x00007f8763010094 in Py_InitializeEx ()
       from /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
    #6  0x00007f876347c645 in init (lang=<optimized out>) at python.c:375
    #7  0x00007f87641a805e in r_lang_add (lang=0x557e87ad4840, 
        foo=0x7f8763681280 <r_lang_plugin_python2>) at lang.c:133
    #8  0x00007f87687fd0e5 in __lib_lang_cb (pl=0x557e87bc61e0, user=0x557e85ca3540 <r>, 
        data=0x7f8763681280 <r_lang_plugin_python2>) at libs.c:37
    #9  0x00007f8764a07270 in r_lib_run_handler (lib=0x557e87be1310, plugin=0x557e87bc61e0, 
        symbol=0x7f8763681860 <radare_plugin>) at lib.c:171
    #10 0x00007f8764a07998 in r_lib_open_ptr (lib=0x557e87be1310, 
        file=0x7ffd6ba0e9b0 "/home/martin/.config/radare2/plugins/lang_python2.so", 
        handler=0x557e87bc4da0, stru=0x7f8763681860 <radare_plugin>) at lib.c:320
    #11 0x00007f8764a077b2 in r_lib_open (lib=0x557e87be1310, 
        file=0x7ffd6ba0e9b0 "/home/martin/.config/radare2/plugins/lang_python2.so") at lib.c:280
    #12 0x00007f8764a07b76 in r_lib_opendir (lib=0x557e87be1310, 
        path=0x557e87bd0320 "/home/martin/.config/radare2/plugins") at lib.c:400
    #13 0x00007f87687fd746 in r_core_loadlibs (core=0x557e85ca3540 <r>, where=-1, path=0x0)
        at libs.c:91
    #14 0x0000557e85a9de30 in main (argc=3, argv=0x7ffd6ba0f238, envp=0x7ffd6ba0f258)
        at radare2.c:750
    (gdb)

### PR DESCRIPTION
**Pull request by mcarpenter**
Resolves #8894

### CODE REVIEW

**Comment by mcarpenter** [comment]
Meh, that was a little trigger happy (sorry! end of the day): should also check the return code from `dup2()`.

**Comment by mcarpenter** [comment]
Okay, I think I'm done.

(Practically speaking I think it's near impossible to hit the `dup2()` branch (given that POSIX mandates that `open()` returns the lowest FD) but this should be robust if there is some weird corner case where that might happen).

File: binr/radare2/radare2.c [hunk 1]
```diff
@@ -658,7 +658,24 @@ int main(int argc, char **argv, char **envp) {
 		}
 	}
 	if (noStderr) {
-		close (2);
+		if (-1 == close (2)) {
+			eprintf ("Failed to close stderr");
+			return 1;
+		}
+		int new_stderr = open ("/dev/null", O_RDWR);
```
Comment (radare) [inline binr/radare2/radare2.c]: that will not work on a system without /dev mounted, also, this is wrong for windows

Comment (radare) [reply binr/radare2/radare2.c]: but it's a good fix for that issue

File: binr/radare2/radare2.c [hunk 2]
```diff
@@ -658,7 +658,24 @@ int main(int argc, char **argv, char **envp) {
...
+		if (-1 == new_stderr) {
+			eprintf ("Failed to open /dev/null");
+			return 1;
+		} else if (2 != new_stderr) {
```
Comment (radare) [inline binr/radare2/radare2.c]: don't else after return

**Comment by mcarpenter** [comment]
I added the `null` vs. `/dev/null` idiom found in [libr/bin/pdb/pdb_downloader.c](https://github.com/radare/radare2/blob/d909c3d3335fb184a0f5ee6c3ee5df0296960279/libr/bin/pdb/pdb_downloader.c#L22) and made the style change requested.

Systems that don't have `/dev` or `/dev/null` aren't [POSIX compliant](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap10.html) ("shall exist on conforming systems") so I'm not sure that's an issue. Did you have a particular case in mind?

**Comment by radare** [comment]
Thanks!

**Comment by mcarpenter** [comment]
No, thank you!

**Pull request merged**